### PR TITLE
Issue/#320 update geolite database

### LINKED ISSUE #320: Server is sending empty strings as countries for some players
This is happening since last server migration.

### PR DESCRIPTION
**Pull request by Askaholic**
Fixes #320 

Add service for updating the geoip database automatically. GeoLite2 releases a new version every month, so I set the default max age to 32 days.

With these changes we should be able to remove the database from git. Let me know if I should add that to this PR.

`LobbyConnection.db` is never used so I removed it

### CODE REVIEW

File: server/__init__.py [hunk 1]
```diff
@@ -53,6 +54,7 @@
 
 
 def run_lobby_server(address: (str, int),
+                     geoip_service: GeoIpService,
```
Comment (norraxx) [inline server/__init__.py]: If you do something like that, add parameter at the end of function

File: server/geoip_service.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import shutil
+
+from .decorators import with_logger
+from . import config
+
+import traceback
+
+
+@with_logger
+class GeoIpService(object):
```
Comment (norraxx) [inline server/geoip_service.py]: You write code for peole, please write comments

File: server/geoip_service.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        # Unzip the archive and overwrite the old file
+        try:
+            with gzip.open(temp_file_path, 'rb') as f_in:
+                with open(self.file_path, 'wb') as f_out:
+                    shutil.copyfileobj(f_in, f_out)
+        except OSError:  # pragma: no cover
+            self._logger.warning("Failed to unzip downloaded file!")
+        self._logger.info("New database download complete")
+
+    async def _download_file(self, url: str, file_path: str) -> None:
```
Comment (norraxx) [inline server/geoip_service.py]: `url` is not used

File: server/geoip_service.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    the service will try to download a new file from tue url in ``server.config``.
+
+        Provides an interface for getting data out of the database.
+    """
+
+    def __init__(self):
+        self.file_path = config.GEO_IP_DATABASE_PATH
+
+        # crontab: min hour day month day_of_week
+        self._update_cron = aiocron.crontab('*/10 * * * *', func=self.check_update_geoip_db)
```
Comment (norraxx) [inline server/geoip_service.py]: let's check once per day?

File: server/geoip_service.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+from datetime import datetime
+import gzip
+import geoip2.database
+import os
+import shutil
+
+from .decorators import with_logger
+from . import config
+
+import traceback
```
Comment (norraxx) [inline server/geoip_service.py]: import asyncio
import gzip
import os
import shutil
import traceback
from datetime import datetime

import aiocron
import aiohttp
import geoip2.database

from . import config
from .decorators import with_logger

File: tests/unit_tests/test_geoip_service.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+from aiohttp import web
+import gzip
+import os
+import random
+import string
+from unittest.mock import Mock
+
+import server.config
+from server.geoip_service import GeoIpService
```
Comment (norraxx) [inline tests/unit_tests/test_geoip_service.py]: import gzip
import os
import random
import string
from unittest.mock import Mock

from aiohttp import web

import server.config
from server.geoip_service import GeoIpService

File: tests/unit_tests/test_geoip_service.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+
+async def test_check_update(geoip_service):
+    path = '/tmp/fake_db.mmdb'
+    with open(path, 'a'):
+        os.utime(path, (0, 0))
+    server.config.GEO_IP_DATABASE_MAX_AGE_DAYS = 32
+    geoip_service.file_path = '/tmp/fake_db.mmdb'
```
Comment (norraxx) [inline tests/unit_tests/test_geoip_service.py]: geoip_service.file_path = '/tmp/fake_db.mmdb' - copy paste kung-fu, how about to move all common configuration to fixure?

For different cases: how about to parametrize fixture via `request`? https://docs.pytest.org/en/latest/example/parametrize.html

File: tests/unit_tests/test_geoip_service.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    await loop.create_server(app.make_handler(), '127.0.0.1', PORT)
+
+    # Test geoip update
+    if os.path.isfile(path):
+        os.remove(path)
+    # Need to call it like this because the function is mocked
+    await GeoIpService.download_geoip_db(geoip_service)
+    assert os.path.isfile(path) is True
+    with open(path) as f:
+        assert f.read() == random_text
```
Comment (norraxx) [inline tests/unit_tests/test_geoip_service.py]: do you clean files on local before + after test?

File: tests/unit_tests/test_geoip_service.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    path = '/tmp/fake_db.mmdb'
+    with open(path, 'a'):
+        os.utime(path, (0, 0))
+    server.config.GEO_IP_DATABASE_MAX_AGE_DAYS = 32
+    geoip_service.file_path = '/tmp/fake_db.mmdb'
+    geoip_service.load_db = Mock()
+
+    await geoip_service.check_update_geoip_db()
+
+    geoip_service.download_geoip_db.assert_called_once_with()
```
Comment (norraxx) [inline tests/unit_tests/test_geoip_service.py]: assert_called_once ?

Comment (Askaholic) [reply tests/unit_tests/test_geoip_service.py]: Doesn't exist in python3.5. Need 3.6 for that :P

https://docs.python.org/3.5/library/unittest.mock.html#unittest.mock.Mock.assert_called_once_with

File: tests/unit_tests/test_geoip_service.py (hunk 8 above)
Comment (Askaholic) [reply tests/unit_tests/test_geoip_service.py]: Yes, look on line 48

**Comment by Rackover** [comment]
```player_id =  %s with args: 1
ERROR    server.lobbyconnection.LobbyConnection The country method cannot be used with the GeoLite2-City database
Traceback (most recent call last):
  File "/home/rackover/share/server/server/lobbyconnection.py", line 126, in on_message_received
    await handler(message)
  File "/home/rackover/share/server/server/lobbyconnection.py", line 564, in command_hello
    self.player.country = self.geoip_service.country(self.peer_address.host)
  File "/home/rackover/share/server/server/geoip_service.py", line 118, in country
    return str(self.db.country(address).country.iso_code)
  File "/home/rackover/.local/lib/python3.6/site-packages/geoip2/database.py", line 104, in country
    return self._model_for(geoip2.models.Country, 'Country', ip_address)
  File "/home/rackover/.local/lib/python3.6/site-packages/geoip2/database.py", line 195, in _model_for
    record = self._get(types, ip_address)
  File "/home/rackover/.local/lib/python3.6/site-packages/geoip2/database.py", line 187, in _get
    self.metadata().database_type))
TypeError: The country method cannot be used with the GeoLite2-City database
WARNING  server.lobbyconnection.LobbyConnection Client Gardelouve dropped. Error processing command
DEBUG    server.servercontext.ServerContext ServerContext(LobbyServer): Client connected
DEBUG    server.lobbyconnection.LobbyConnection LobbyConnection initialized
```

**Comment by Askaholic** [comment]
Aaah. I put the wrong database url in the config. Should be the country db not the city one.

**Pull request merged**